File: guilib-bukkit/src/main/java/net/developertobi/guilib/bukkit/api/GuiApiImpl.java
package net.developertobi.guilib.bukkit.api;

import net.developertobi.guilib.api.GuiApi;
import net.developertobi.guilib.api.GuiItemAction;
import net.developertobi.guilib.api.gui.GuiHandler;
import net.developertobi.guilib.api.item.GuiItem;
import net.developertobi.guilib.api.item.GuiPos;
import net.developertobi.guilib.api.pagination.GuiPagination;
import net.developertobi.guilib.api.GuiHelper;
import net.developertobi.guilib.bukkit.api.gui.ConfirmationGui;
import net.developertobi.guilib.bukkit.api.gui.GuiHandlerImpl;
import net.developertobi.guilib.bukkit.api.item.GuiItemImpl;
import net.kyori.adventure.text.Component;
import org.bukkit.Material;
import org.bukkit.NamespacedKey;
import org.bukkit.entity.Player;
import org.bukkit.event.inventory.InventoryClickEvent;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;
import org.bukkit.persistence.PersistentDataContainer;
import org.bukkit.persistence.PersistentDataType;

import java.util.UUID;

public class GuiApiImpl implements GuiApi {

    private final GuiHandler guiHandler = new GuiHandlerImpl();

    @Override
    public GuiHandler getGuiHandler() {
        return guiHandler;
    }

    @Override
    public GuiItem placeholder(Material type) {
        String itemId = UUID.randomUUID().toString().split("-")[0];
        return of(makeItemStack(type, itemId));
    }

    @Override
    public GuiItem nextPage(ItemStack item, GuiPagination pagination) {
        return of(item, (pos, guiItem, event) -> pagination.toNextPage());
    }

    @Override
    public GuiItem previousPage(ItemStack item, GuiPagination pagination) {
        return of(item, (pos, guiItem, event) -> pagination.toPreviousPage());
    }

    @Override
    public GuiItem navigator(ItemStack item, String guiKey) {
        return of(item, (pos, guiItem, event) -> {
            Player holder = (Player) event.getWhoClicked();
            var gui = GuiHelper.getGui(holder, guiKey);
            if (gui == null) return;

            holder.closeInventory();
            gui.open(holder);
        });
    }

    @Override
    public GuiItem of(ItemStack item) {
        return new GuiItemImpl(item);
    }

    @Override
    public GuiItem of(ItemStack item, GuiItemAction action) {
        return new GuiItemImpl(item, action);
    }

    @Override
    public void openConfirmationGui(Player holder, Component title, ItemStack displayItem,
                                          java.util.function.Consumer<ItemStack> onAccept,
                                          java.util.function.Consumer<ItemStack> onDeny) {
        guiHandler.openStaticView(
                holder,
                title,
                new ConfirmationGui(displayItem, onAccept, onDeny),
                true
        );
    }

    private ItemStack makeItemStack(Material type, String itemId) {
        ItemStack itemStack = new ItemStack(type);
        ItemMeta itemMeta = itemStack.getItemMeta();
        if (itemMeta != null) {
            itemMeta.displayName(Component.text(" "));

            NamespacedKey namespacedKey = new NamespacedKey("item", "itemid");
            PersistentDataContainer dataContainer = itemMeta.getPersistentDataContainer();
            if (!dataContainer.has(namespacedKey, PersistentDataType.STRING)) {
                dataContainer.set(namespacedKey, PersistentDataType.STRING, itemId);
            }
            itemStack.setItemMeta(itemMeta);
        }
        return itemStack;
    }
}

File: guilib-bukkit/bin/main/net/developertobi/inventorylib/bukkit/api/GuiApiImpl.kt
package net.developertobi.inventorylib.bukkit.api

import net.kyori.adventure.text.Component
import net.developertobi.inventorylib.api.GuiApi
import net.developertobi.inventorylib.api.extension.getInventory
import net.developertobi.inventorylib.api.inventory.GuiHandler
import net.developertobi.inventorylib.api.item.GuiItem
import net.developertobi.inventorylib.api.item.GuiPos
import net.developertobi.inventorylib.api.pagination.GuiPagination
import net.developertobi.inventorylib.bukkit.api.inventory.ConfirmationGui
import net.developertobi.inventorylib.bukkit.api.inventory.GuiHandlerImpl
import net.developertobi.inventorylib.bukkit.api.item.GuiItemImpl
import org.bukkit.Material
import org.bukkit.NamespacedKey
import org.bukkit.entity.Player
import org.bukkit.event.inventory.InventoryClickEvent
import org.bukkit.inventory.ItemStack
import org.bukkit.inventory.meta.ItemMeta
import org.bukkit.persistence.PersistentDataContainer
import org.bukkit.persistence.PersistentDataType
import java.util.*

class GuiApiImpl : GuiApi {

    override val inventoryHandler: GuiHandler = GuiHandlerImpl()

    override fun placeholder(type: Material): GuiItem {
        val itemId = UUID.randomUUID().toString().split("-")[0]
        return of(makeItemStack(type, itemId)) { _, _, _ -> }
    }

    override fun nextPage(item: ItemStack, pagination: GuiPagination): GuiItem {
        return of(item) { _: GuiPos, _: GuiItem, _: InventoryClickEvent? -> pagination.toNextPage() }
    }

    override fun previousPage(item: ItemStack, pagination: GuiPagination): GuiItem {
        return of(item) { _: GuiPos, _: GuiItem, _: InventoryClickEvent -> pagination.toPreviousPage() }
    }

    override fun navigator(item: ItemStack, inventoryKey: String): GuiItem {
        return of(item) { _: GuiPos, _: GuiItem, event: InventoryClickEvent ->
            val holder: Player = event.whoClicked as Player
            val inventory = getInventory(holder, inventoryKey) ?: return@of

            holder.closeInventory()
            inventory.open(holder)
        }
    }

    override fun of(item: ItemStack): GuiItem {
        return GuiItemImpl(item) { _: GuiPos, _: GuiItem, _: InventoryClickEvent -> }
    }

    override fun of(item: ItemStack, action: ((GuiPos, GuiItem, InventoryClickEvent) -> Unit)): GuiItem {
        return GuiItemImpl(item, action)
    }

    override fun openConfirmationInventory(holder: Player, title: Component, displayItem: ItemStack, onAccept: ((ItemStack) -> Unit), onDeny: ((ItemStack) -> Unit)) {
        inventoryHandler.openStaticView(
            holder = holder,
            title = title,
            provider = ConfirmationGui(displayItem, onAccept, onDeny),
            forceSyncOpening = true
        )
    }

    private fun makeItemStack(type: Material, itemId: String): ItemStack {
        val itemStack = ItemStack(type)
        val itemMeta: ItemMeta = itemStack.itemMeta
        itemMeta.displayName(Component.text(" "))

        val namespacedKey = NamespacedKey("item", "itemid")
        val dataContainer: PersistentDataContainer = itemMeta.persistentDataContainer
        if (!dataContainer.has(namespacedKey, PersistentDataType.STRING)) {
            dataContainer.set(namespacedKey, PersistentDataType.STRING, itemId)
            itemStack.setItemMeta(itemMeta)
        }

        itemStack.setItemMeta(itemMeta)
        return itemStack
    }
}

